File: src/lib/generate_excel.py
import sys
import json
import openpyxl
import copy
import datetime
import tempfile
import os
from openpyxl.styles import Font, Alignment, PatternFill, Border, Side

def format_date_id(date_str):
    months_id = {
        "01": "Jan", "02": "Feb", "03": "Mar", "04": "Apr", "05": "Mei", "06": "Jun",
        "07": "Jul", "08": "Agu", "09": "Sep", "10": "Okt", "11": "Nov", "12": "Des"
    }
    try:
        parts = date_str.split("-")
        day = str(int(parts[2]))
        month = months_id[parts[1]]
        year = parts[0][2:]
        return f"{day} {month} {year}"
    except:
        return date_str

def parse_time(time_str):
    try:
        parts = time_str.split(":")
        return datetime.time(int(parts[0]), int(parts[1]))
    except:
        return time_str

def is_dosen_luar(row):
    # Sumber kebenaran: field "status" (dari kolom status_dosen di database, diatur
    # lewat halaman List Dosen). Daftar kata kunci di bawah cuma fallback untuk baris
    # lama/offline yang belum membawa field status sama sekali.
    status = row.get("status")
    if status:
        return status == "luar"

    name_lower = row.get("dsn", "").lower()
    luar_keywords = [
        "anggi sandika",
        "adityana",
        "amelia",
        "hendy satria",
        "handy satria",
        "i'oh",
        "iin nurhayati",
        "sonny feisal",
        "sussylawati",
        "tarma"
    ]
    return any(keyword in name_lower for keyword in luar_keywords)

def main():
    try:
        # Baca JSON payload dari stdin
        payload = json.loads(sys.stdin.read())
        
        # Path dihitung relatif terhadap lokasi skrip ini (src/lib/generate_excel.py),
        # supaya tidak bergantung pada drive/folder proyek di komputer tertentu.
        project_root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
        template_path = os.path.join(project_root, "File Pendukung", "Honjar Mei Tahun 2026 PRODI TLM D4.xlsx")
        if not os.path.exists(template_path):
            print(f"ERROR: Template file not found at {template_path}", file=sys.stderr)
            sys.exit(1)
            
        wb = openpyxl.load_workbook(template_path)
        
        month_name = payload.get("monthName", "Mei 2026").upper()
        current_date_str = payload.get("currentDate", "8 Juni 2026")
        rows_data = payload.get("rows", [])
        
        # Kelompokkan ke Dosen Tetap dan Dosen Luar
        rows_tetap = []
        rows_luar = []
        
        for r in rows_data:
            if is_dosen_luar(r):
                rows_luar.append(r)
            else:
                rows_tetap.append(r)
                
        # Urutkan: Dosen -> Tanggal -> Jam Mulai
        rows_tetap.sort(key=lambda x: (x["dsn"], x["tgl"], x["a"]))
        rows_luar.sort(key=lambda x: (x["dsn"], x["tgl"], x["a"]))
        
        # Eksekusi untuk masing-masing Sheet
        for sheet_name, data in [("Tetap", rows_tetap), ("Luar", rows_luar)]:
            sheet = wb[sheet_name]
            
            # Ganti periode pada judul di baris 3
            sheet.cell(row=3, column=1).value = f"PERIODE {month_name}"
            
            # Ubah jenis font untuk semua cell di baris 1 s/d 6 menjadi Times New Roman
            for r_idx in range(1, 7):
                for col_idx in range(1, 16):
                    cell = sheet.cell(row=r_idx, column=col_idx)
                    if cell.font:
                        f_copy = copy.copy(cell.font)
                        f_copy.name = "Times New Roman"
                        cell.font = f_copy
            
            # Ambil style formatting dasar dari baris data ke-7 template
            template_styles = []
            for col in range(1, 12):
                cell = sheet.cell(row=7, column=col)
                template_styles.append({
                    'font': copy.copy(cell.font),
                    'alignment': copy.copy(cell.alignment),
                    'fill': copy.copy(cell.fill),
                    'border': copy.copy(cell.border),
                    'number_format': cell.number_format
                })

            # Tinggi baris data normal, diambil dari baris pertama yang tidak dipakai
            # sebagai baris style (baris 8 pada template asli). Dipakai untuk menimpa
            # tinggi baris 7 bawaan template yang sengaja dibuat sangat kecil (0.95px,
            # baris bantu kosong) — kalau tidak ditimpa, baris data pertama ikut kecil.
            normal_row_height = sheet.row_dimensions[8].height if 8 in sheet.row_dimensions else 15

            # Isian baris pemisah kuning antar kelompok dosen
            yellow_fill = PatternFill(start_color="FFFFFF00", end_color="FFFFFF00", fill_type="solid")

            # Bersihkan seluruh data lama dari baris 7 sampai baris terakhir secara efisien
            sheet.delete_rows(7, sheet.max_row - 6)

            # Tulis data baru mulai baris 7
            current_dsn = None
            no_counter = 0

            start_row = 7
            row_num = start_row
            for r in data:
                if r["dsn"] != current_dsn:
                    # Baris pemisah kuning sebelum kelompok dosen baru (bukan sebelum yang pertama)
                    if current_dsn is not None:
                        sheet.row_dimensions[row_num].height = normal_row_height
                        for col_idx in range(1, 12):
                            cell = sheet.cell(row=row_num, column=col_idx)
                            cell.value = None
                            style = template_styles[col_idx - 1]
                            if style['border']: cell.border = copy.copy(style['border'])
                            cell.fill = yellow_fill
                        row_num += 1

                    no_counter += 1
                    no_val = no_counter
                    current_dsn = r["dsn"]
                else:
                    no_val = None

                # Tulis data serta formula Excel
                cell_vals = {
                    1: no_val,                           # NO
                    2: r["dsn"],                         # DOSEN
                    3: format_date_id(r["tgl"]),         # TANGGAL
                    4: parse_time(r["a"]),               # MULAI
                    5: parse_time(r["b"]),               # SELESAI
                    6: f"=VALUE(K{row_num})",            # DURASI
                    7: f"=IF(F{row_num}>=300,6,IF(F{row_num}>=250,5,IF(F{row_num}>=200,4,IF(F{row_num}>=150,3,IF(F{row_num}>=100,2,IF(F{row_num}>=50,1,0))))))", # JUMLAH JAM
                    8: r["mk"],                          # MATA KULIAH
                    9: r["met"],                         # METODE
                    10: r["kls"],                        # KELAS
                    11: f'=TEXT(E{row_num}-D{row_num},"[mm]")' # Hitung Menit (Helper)
                }

                sheet.row_dimensions[row_num].height = normal_row_height
                for col_idx in range(1, 12):
                    cell = sheet.cell(row=row_num, column=col_idx)
                    cell.value = cell_vals[col_idx]

                    # Salin style template dan ubah font name ke Times New Roman
                    style = template_styles[col_idx - 1]
                    if style['font']:
                        f_copy = copy.copy(style['font'])
                        f_copy.name = "Times New Roman"
                        cell.font = f_copy
                    if style['alignment']: cell.alignment = copy.copy(style['alignment'])
                    if style['fill']: cell.fill = copy.copy(style['fill'])
                    if style['border']: cell.border = copy.copy(style['border'])
                    if style['number_format']: cell.number_format = style['number_format']

                row_num += 1

            # Tulis Tanda Tangan Ka. Prodi di akhir tabel (3 baris setelah data berakhir)
            last_data_row = row_num - 1
            if len(data) == 0:
                last_data_row = 6

            sig_start = last_data_row + 4
            
            sheet.cell(row=sig_start, column=8).value = f"Cimahi, {current_date_str}"
            sheet.cell(row=sig_start, column=8).font = Font(name="Times New Roman", size=10)
            
            sheet.cell(row=sig_start + 1, column=8).value = "Ka. Prodi TLM D4"
            sheet.cell(row=sig_start + 1, column=8).font = Font(name="Times New Roman", size=10, bold=True)
            
            sheet.cell(row=sig_start + 5, column=8).value = "Gina Khairinisa,M.Imun"
            sheet.cell(row=sig_start + 5, column=8).font = Font(name="Times New Roman", size=10, bold=True)
            
            sheet.cell(row=sig_start + 6, column=8).value = "NID. 4121 282 88"
            sheet.cell(row=sig_start + 6, column=8).font = Font(name="Times New Roman", size=10)
            
        # Simpan ke temp file
        fd, temp_file_path = tempfile.mkstemp(suffix=".xlsx")
        os.close(fd)
        wb.save(temp_file_path)
        
        # Cetak output path ke stdout agar dibaca Next.js
        print(temp_file_path)
        
    except Exception as e:
        print(f"ERROR: {str(e)}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()

File: src/lib/generateExcel.ts
import ExcelJS from "exceljs";
import fs from "fs";
import path from "path";

const MONTHS = [
  "Jan", "Feb", "Mar", "Apr", "Mei", "Jun",
  "Jul", "Agu", "Sep", "Okt", "Nov", "Des",
];

const LUAR_KEYWORDS = [
  "anggi sandika",
  "adityana",
  "amelia",
  "hendy satria",
  "handy satria",
  "i'oh",
  "iin nurhayati",
  "sonny feisal",
  "sussylawati",
  "tarma",
];

function formatTanggal(tgl?: string): string {
  if (!tgl) return "";

  const d = new Date(tgl);
  if (Number.isNaN(d.getTime())) return tgl;

  return `${d.getDate()} ${MONTHS[d.getMonth()]} ${String(d.getFullYear()).slice(-2)}`;
}

function formatDateId(dateStr?: string): string {
  if (!dateStr) return "";

  const d = new Date(dateStr);
  if (Number.isNaN(d.getTime())) return dateStr;

  const day = d.getDate();
  const month = MONTHS[d.getMonth()];
  const year = String(d.getFullYear()).slice(-2);
  return `${day} ${month} ${year}`;
}

function formatTimeValue(value: unknown): string {
  if (value === null || value === undefined || value === "") return "";

  if (typeof value === "string") {
    const trimmed = value.trim();
    if (/^\d{1,2}:\d{2}$/.test(trimmed)) return trimmed;
    if (/^\d{1,2}:\d{2}:\d{2}$/.test(trimmed)) {
      const [h, m] = trimmed.split(":");
      return `${String(Number(h)).padStart(2, "0")}:${String(Number(m)).padStart(2, "0")}`;
    }
    if (/^\d{4}-\d{2}-\d{2}T/.test(trimmed) || !Number.isNaN(Date.parse(trimmed))) {
      const d = new Date(trimmed);
      if (!Number.isNaN(d.getTime())) {
        return `${String(d.getHours()).padStart(2, "0")}:${String(d.getMinutes()).padStart(2, "0")}`;
      }
    }
    return trimmed;
  }

  if (value instanceof Date) {
    return `${String(value.getHours()).padStart(2, "0")}:${String(value.getMinutes()).padStart(2, "0")}`;
  }

  if (typeof value === "number") {
    const totalMinutes = Math.round((value % 1) * 24 * 60);
    const hours = Math.floor(totalMinutes / 60) % 24;
    const minutes = totalMinutes % 60;
    return `${String(hours).padStart(2, "0")}:${String(minutes).padStart(2, "0")}`;
  }

  return String(value).trim();
}

function normalizeStatus(row: any): "tetap" | "luar" {
  if (row?.status === "luar" || row?.status === "tetap") {
    return row.status;
  }

  const source = String(row?.dsn ?? row?.dosen ?? row?.nama ?? "").toLowerCase();
  return LUAR_KEYWORDS.some((keyword) => source.includes(keyword)) ? "luar" : "tetap";
}

function sortRows(rows: any[]) {
  return [...rows].sort((a, b) => {
    const left = `${a.dsn ?? ""}|${a.tgl ?? ""}|${a.a ?? ""}`.toLowerCase();
    const right = `${b.dsn ?? ""}|${b.tgl ?? ""}|${b.a ?? ""}`.toLowerCase();
    return left.localeCompare(right);
  });
}

function clearRows(worksheet: ExcelJS.Worksheet, startRow: number) {
  const lastRow = worksheet.rowCount || 400;
  for (let row = startRow; row <= lastRow; row += 1) {
    for (let col = 1; col <= 12; col += 1) {
      worksheet.getCell(row, col).value = null;
      worksheet.getCell(row, col).font = { name: "Times New Roman", size: 10 };
      worksheet.getCell(row, col).alignment = { vertical: "middle" };
    }
  }
}

function setBorder(cell: ExcelJS.Cell) {
  cell.border = {
    top: { style: "thin", color: { argb: "FF000000" } },
    left: { style: "thin", color: { argb: "FF000000" } },
    bottom: { style: "thin", color: { argb: "FF000000" } },
    right: { style: "thin", color: { argb: "FF000000" } },
  };
}

function setHeaderCell(worksheet: ExcelJS.Worksheet, row: number, col: number, value: string) {
  const cell = worksheet.getCell(row, col);
  cell.value = value;
  cell.font = { name: "Times New Roman", size: 10, bold: true };
  cell.alignment = { horizontal: "center", vertical: "middle" };
  cell.fill = { type: "pattern", pattern: "solid", fgColor: { argb: "FFF2F2F2" } };
  setBorder(cell);
}

function writeGroupedSheet(
  worksheet: ExcelJS.Worksheet,
  rows: any[],
  monthName: string,
  currentDate: string,
  sheetName: "Tetap" | "Luar"
) {
  worksheet.name = sheetName;
  worksheet.views = [{ state: "normal" }];

  // 1. LEBAR KOLOM SESUAI PROPORSI GAMBAR
  worksheet.columns = [
    { width: 5 },   // 1. NO
    { width: 32 },  // 2. DOSEN TETAP / DOSEN LUAR (Lebar)
    { width: 14 },  // 3. TANGGAL
    { width: 10 },  // 4. MULAI
    { width: 10 },  // 5. SELESAI
    { width: 10 },  // 6. DURASI
    { width: 10 },  // 7. JUMLAH JAM
    { width: 35 },  // 8. MATA KULIAH (Lebar)
    { width: 14 },  // 9. METODE
    { width: 18 },  // 10. TINGKAT / KELAS
    { width: 12 },  // 11. Hitung Menit
  ];

  // Header Judul Dokumen
  worksheet.getCell("A1").value = "REKAPITULASI JAM MENGAJAR DOSEN PROGRAM STUDI TEKNOLOGI LABORATORIUM MEDIK (D4)";
  worksheet.getCell("A1").font = { name: "Times New Roman", size: 12, bold: true };
  worksheet.getCell("A1").alignment = { horizontal: "center", vertical: "middle" };
  worksheet.mergeCells("A1:K1");

  worksheet.getCell("A2").value = `SEMESTER GENAP TA. 2025/2026`;
  worksheet.getCell("A2").font = { name: "Times New Roman", size: 10, bold: true };
  worksheet.getCell("A2").alignment = { horizontal: "center", vertical: "middle" };
  worksheet.mergeCells("A2:K2");

  worksheet.getCell("A3").value = `PERIODE ${monthName.toUpperCase()}`;
  worksheet.getCell("A3").font = { name: "Times New Roman", size: 10, bold: true };
  worksheet.getCell("A3").alignment = { horizontal: "center", vertical: "middle" };
  worksheet.mergeCells("A3:K3");

  // 2. SUSUN HEADER BERTINGKAT 2 BARIS (BARIS 5 & 6)
  // Set nilai header dasar
  setHeaderCell(worksheet, 5, 1, "NO");
  setHeaderCell(worksheet, 5, 2, sheetName === "Tetap" ? "DOSEN TETAP" : "DOSEN LUAR");
  setHeaderCell(worksheet, 5, 3, "TANGGAL");
  setHeaderCell(worksheet, 5, 4, "WAKTU PERKULIAHAN"); // Merged D5:E5
  setHeaderCell(worksheet, 6, 4, "MULAI");
  setHeaderCell(worksheet, 6, 5, "SELESAI");
  setHeaderCell(worksheet, 5, 6, "DURASI");
  setHeaderCell(worksheet, 5, 7, "JUMLAH\nJAM");
  setHeaderCell(worksheet, 5, 8, "MATA KULIAH");
  setHeaderCell(worksheet, 5, 9, "METODE");
  setHeaderCell(worksheet, 5, 10, "TINGKAT / KELAS");
  setHeaderCell(worksheet, 5, 11, "Hitung Menit");

  // Set style header baris 6 untuk Mulai & Selesai
  setHeaderCell(worksheet, 6, 4, "MULAI");
  setHeaderCell(worksheet, 6, 5, "SELESAI");

  // Merge Cell Header Vertical (Baris 5-6) untuk kolom selain Waktu Perkuliahan
  worksheet.mergeCells("A5:A6");
  worksheet.mergeCells("B5:B6");
  worksheet.mergeCells("C5:C6");
  worksheet.mergeCells("D5:E5"); // Horizontal Merge untuk WAKTU PERKULIAHAN
  worksheet.mergeCells("F5:F6");
  worksheet.mergeCells("G5:G6");
  worksheet.mergeCells("H5:H6");
  worksheet.mergeCells("I5:I6");
  worksheet.mergeCells("J5:J6");
  worksheet.mergeCells("K5:K6");

  // Pastikan border tetap rapi di sel yang di-merge
  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11].forEach((col) => {
    setBorder(worksheet.getCell(5, col));
    setBorder(worksheet.getCell(6, col));
    worksheet.getCell(5, col).alignment = { horizontal: "center", vertical: "middle", wrapText: true };
    worksheet.getCell(6, col).alignment = { horizontal: "center", vertical: "middle", wrapText: true };
  });

  worksheet.getRow(5).height = 22;
  worksheet.getRow(6).height = 18;

  // 3. PENGISIAN DATA (DIMULAI DARI BARIS 7)
  let rowNum = 7;
  let currentDosen: string | null = null;
  let noCounter = 0;

  rows.forEach((row) => {
    const dosen = String(row?.dsn ?? row?.dosen ?? "").trim();
    const isNewDosen = Boolean(dosen) && dosen !== currentDosen;

    if (isNewDosen) {
      noCounter += 1;
      currentDosen = dosen;

      if (rowNum > 7) {
        // Pemisah antar dosen (Baris Kuning/Kosong)
        for (let c = 1; c <= 11; c++) {
          worksheet.getCell(rowNum, c).fill = { type: "pattern", pattern: "solid", fgColor: { argb: "FFFFFF00" } };
        }
        rowNum += 1;
      }
    }

    const noVal = isNewDosen ? noCounter : null;
    const startTime = formatTimeValue(row?.a ?? row?.mulai ?? row?.jam_mulai ?? row?.startTime);
    const endTime = formatTimeValue(row?.b ?? row?.selesai ?? row?.jam_selesai ?? row?.endTime);
    const durationMinutes = Number(row?.mnt ?? 0) || ((startTime && endTime) ? (() => {
      const [sh, sm] = startTime.split(":").map(Number);
      const [eh, em] = endTime.split(":").map(Number);
      return Math.max(0, (eh * 60 + em) - (sh * 60 + sm));
    })() : 0);

    const rowData = [
      noVal,
      dosen,
      formatDateId(row?.tgl),
      startTime,
      endTime,
      durationMinutes,
      Number(row?.jam ?? 0) || Math.max(1, Math.ceil(durationMinutes / 100)),
      row?.mk || "",
      row?.met || "",
      row?.kls || "",
      durationMinutes,
    ];

    rowData.forEach((value, idx) => {
      const cell = worksheet.getCell(rowNum, idx + 1);
      cell.value = value ?? "";
      cell.font = { name: "Times New Roman", size: 10 };
      // Align kiri untuk Dosen, Mata Kuliah, dan Kelas
      cell.alignment = { 
        vertical: "middle", 
        horizontal: (idx === 1 || idx === 7 || idx === 9) ? "left" : "center" 
      };
      setBorder(cell);
      if (idx === 0 && noVal !== null) {
        cell.fill = { type: "pattern", pattern: "solid", fgColor: { argb: "FFF7F7F7" } };
      }
    });

    worksheet.getCell(rowNum, 2).font = { name: "Times New Roman", size: 10, bold: true };
    worksheet.getRow(rowNum).height = 20;
    rowNum += 1;
  });

  // Tanda Tangan Ka. Prodi
  const signatureRow = rowNum + 2;
  worksheet.getCell(signatureRow, 8).value = `Cimahi, ${currentDate}`;
  worksheet.getCell(signatureRow, 8).font = { name: "Times New Roman", size: 10 };
  worksheet.getCell(signatureRow + 1, 8).value = "Ka. Prodi TLM D4";
  worksheet.getCell(signatureRow + 1, 8).font = { name: "Times New Roman", size: 10, bold: true };
  worksheet.getCell(signatureRow + 5, 8).value = "Gina Khairinisa, M.Imun";
  worksheet.getCell(signatureRow + 5, 8).font = { name: "Times New Roman", size: 10, bold: true };
  worksheet.getCell(signatureRow + 6, 8).value = "NID. 4121 282 88";
  worksheet.getCell(signatureRow + 6, 8).font = { name: "Times New Roman", size: 10 };
}

export async function generateHonorExcel(payload: any): Promise<Buffer> {
  const workbook = new ExcelJS.Workbook();

  const rows: any[] = Array.isArray(payload?.rows) ? payload.rows : [];
  const monthName = String(payload?.monthName || "Semua Periode");
  const currentDate = String(payload?.currentDate || new Date().toISOString().slice(0, 10));

  const grouped = {
    tetap: sortRows(rows.filter((row) => normalizeStatus(row) === "tetap")),
    luar: sortRows(rows.filter((row) => normalizeStatus(row) === "luar")),
  };

  const templatePath = path.join(process.cwd(), "File Pendukung", "Honjar Mei Tahun 2026 PRODI TLM D4.xlsx");
  if (fs.existsSync(templatePath)) {
    const templateWorkbook = new ExcelJS.Workbook();
    await templateWorkbook.xlsx.readFile(templatePath);
    const templateTetap = templateWorkbook.getWorksheet("Tetap");
    const templateLuar = templateWorkbook.getWorksheet("Luar");

    if (templateTetap) {
      const tetapSheet = workbook.addWorksheet("Tetap", templateTetap);
      clearRows(tetapSheet, 7);
      writeGroupedSheet(tetapSheet, grouped.tetap, monthName, currentDate, "Tetap");
    }

    if (templateLuar) {
      const luarSheet = workbook.addWorksheet("Luar", templateLuar);
      clearRows(luarSheet, 7);
      writeGroupedSheet(luarSheet, grouped.luar, monthName, currentDate, "Luar");
    }

    const buffer = await workbook.xlsx.writeBuffer();
    return Buffer.from(buffer);
  }

  const tetapSheet = workbook.addWorksheet("Tetap");
  const luarSheet = workbook.addWorksheet("Luar");

  writeGroupedSheet(tetapSheet, grouped.tetap, monthName, currentDate, "Tetap");
  writeGroupedSheet(luarSheet, grouped.luar, monthName, currentDate, "Luar");

  const buffer = await workbook.xlsx.writeBuffer();
  return Buffer.from(buffer);
}
